### PR TITLE
Electron: Shadowsocks binaries

### PR DESCRIPTION
**Pull request by alalamav**
* Builds Linux and Windows binaries that use Shadowsocks connection handlers.
* Performs connectivity checks, terminating the program with the appropriate (Outline) error code.
* Not yet tested or integrated with the Outline clients.

### CODE REVIEW

File: outline/electron/main.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	tunDNS        *string
+	proxyHost     *string
+	proxyPort     *int
+	proxyPassword *string
+	proxyCipher   *string
+	logLevel      *string
+	dnsFallback   *bool
+	version       *bool
+}
+var version = ""
```
Comment (bemasc) [inline outline/electron/main.go]: I guess this should be nonempty? How are we going to keep this up to date?

We could use an approach like this: https://blog.alexellis.io/inject-build-time-vars-golang/

File: Makefile [hunk 2]
```diff
@@ -25,8 +25,8 @@ ANDROID_OUTLINE_BUILD_CMD="$(ANDROID_BUILD_CMD) $(IMPORT_PATH)/outline/android $
 ANDROID_INTRA_BUILD_CMD="$(ANDROID_BUILD_CMD) $(IMPORT_PATH)/intra $(IMPORT_PATH)/tunnel $(IMPORT_PATH)/tunnel/intra"
 IOS_BUILD_CMD="GO111MODULE=off $(GOBIND) -a -ldflags $(LDFLAGS) -bundleid org.outline.tun2socks -target=ios/arm,ios/arm64 -tags ios -o $(IOS_ARTIFACT) $(IMPORT_PATH)/outline/apple $(IMPORT_PATH)/outline/shadowsocks"
 MACOS_BUILD_CMD="GO111MODULE=off ./tools/$(GOBIND) -a -ldflags $(LDFLAGS) -bundleid org.outline.tun2socks -target=ios/amd64 -tags ios -o $(MACOS_ARTIFACT) $(IMPORT_PATH)/outline/apple $(IMPORT_PATH)/outline/shadowsocks"
-WINDOWS_BUILD_CMD="$(XGOCMD) -ldflags $(XGO_LDFLAGS) -tags $(XGO_BUILD_TAGS)  --targets=windows/386 -dest $(WINDOWS_BUILDDIR) $(TUN2SOCKS_SRC_PATH)/cmd/tun2socks"
-LINUX_BUILD_CMD="$(XGOCMD) -ldflags $(XGO_LDFLAGS) -tags $(XGO_BUILD_TAGS) --targets=linux/amd64 -dest $(LINUX_BUILDDIR) $(TUN2SOCKS_SRC_PATH)/cmd/tun2socks"
+WINDOWS_BUILD_CMD="$(XGOCMD) -ldflags $(XGO_LDFLAGS)  --targets=windows/386 -dest $(WINDOWS_BUILDDIR) $(ELECTRON_PATH)"
```
[the remark this answers is not in the document]
Comment (alalamav) [reply Makefile]: Fixed.

File: outline/electron/main.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+var lwipWriter io.Writer
+
+func main() {
+	args.tunAddr = flag.String("tunAddr", "10.0.85.2", "TUN interface IP address")
+	args.tunGw = flag.String("tunGw", "10.0.85.1", "TUN interface gateway")
+	args.tunMask = flag.String("tunMask", "255.255.255.0", "TUN interface network mask; prefixlen for IPv6")
+	args.tunDNS = flag.String("tunDNS", "1.1.1.1,9.9.9.9", "Comma-separated list of DNS resolvers for the TUN interface (Windows only)")
```
[the remark this answers is not in the document]
Comment (alalamav) [reply outline/electron/main.go]: Added the missing OpenDNS resolver - note that we updated the Cordova clients to use CloudflareDNS and Quad9 when migrating to Shadowsocks go.

File: outline/electron/main.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+			os.Exit(oss.Unexpected)
+		}
+	}()
+
+	log.Infof("tun2socks running...")
+
+	osSignals := make(chan os.Signal, 1)
+	signal.Notify(osSignals, os.Interrupt, os.Kill, syscall.SIGTERM, syscall.SIGHUP)
+	<-osSignals
+}
```
[the remark this answers is not in the document]
Comment (alalamav) [reply outline/electron/main.go]: Done.

File: outline/electron/main.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	if *args.dnsFallback {
+		// UDP connectivity not supported, fall back to DNS over TCP.
+		core.RegisterUDPConnHandler(dnsfallback.NewUDPHandler())
+	} else {
+		core.RegisterUDPConnHandler(
+			shadowsocks.NewUDPHandler(*args.proxyHost, *args.proxyPort, *args.proxyPassword, *args.proxyCipher, udpTimeout))
+	}
+
+	// Configure LWIP stack to receive input data from the TUN device
+	lwipWriter := core.NewLWIPStack().(io.Writer)
```
[the remark this answers is not in the document]
Comment (alalamav) [reply outline/electron/main.go]: Doesn't seem like it, removed.

File: outline/electron/main.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	go func() {
+		_, err := io.CopyBuffer(lwipWriter, tunDevice, make([]byte, mtu))
+		if err != nil {
+			log.Errorf("Failed to write data to network stack")
```
[the remark this answers is not in the document]
Comment (alalamav) [reply outline/electron/main.go]: Done.

File: outline/electron/main.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	if err != nil {
+		log.Errorf("Failed to perform connectivity checks: %v", err)
+		os.Exit(oss.Unexpected)
+	}
+	if !(code == oss.NoError || code == oss.UDPConnectivity) {
+		log.Errorf("Connectivity checks failed with code %v", code)
+		os.Exit(code)
+	}
+	// Override the user flag if UDP is not supported.
+	*args.dnsFallback = *args.dnsFallback || code == oss.UDPConnectivity
```
[the remark this answers is not in the document]
Comment (alalamav) [reply outline/electron/main.go]: Good suggestion.

File: outline/electron/main.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+	// Open TUN device
+	dnsResolvers := strings.Split(*args.tunDNS, ",")
+	tunDevice, err := tun.OpenTunDevice(*args.tunName, *args.tunAddr, *args.tunGw, *args.tunMask, dnsResolvers)
+	if err != nil {
+		log.Errorf("Failed to open TUN device: %v", err)
+		os.Exit(oss.Unexpected)
+	}
+	// Output packets to TUN device
+	core.RegisterOutputFn(func(data []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (alalamav) [reply outline/electron/main.go]: Done.

File: outline/electron/main.go (hunk 1 above)
Comment (alalamav) [reply outline/electron/main.go]: We're using the `-X main.version=[version]` argument of `go`, allowing us to set the value at compile time.

Comment (bemasc) [reply outline/electron/main.go]: OK. In that case, I would make this `var version String` and add a comment indicating how the value is populated.

Comment (alalamav) [reply outline/electron/main.go]: Done.

**Pull request merged**